libraries/eosiolib/contracts/eosio/key_value_singleton.hpp

```diff
@@ -0,0 +1,129 @@
+#include "../../core/eosio/datastream.hpp"
+#include "../../core/eosio/name.hpp"
+
+#include <eosio/to_key.hpp>
+
+namespace detail {
+   constexpr inline size_t max_stack_buffer_size = 512;
+}
+
+namespace eosio {
+   namespace internal_use_do_not_use {
+      extern "C" {
+         __attribute__((eosio_wasm_import))
+         int64_t kv_erase(uint64_t db, uint64_t contract, const char* key, uint32_t key_size);
+
+         __attribute__((eosio_wasm_import))
+         int64_t kv_set(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, const char* value, uint32_t value_size);
+
+         __attribute__((eosio_wasm_import))
+         bool kv_get(uint64_t db, uint64_t contract, const char* key, uint32_t key_size, uint32_t& value_size);
+
+         __attribute__((eosio_wasm_import))
+         uint32_t kv_get_data(uint64_t db, uint32_t offset, char* data, uint32_t data_size);
+      }
+   }
+
+   template <typename T, eosio::name::raw SingletonName, eosio::name::raw DbName = "eosio.kvram"_n>
```

Comment:
By defining a default parameter here we simplify cases where the type and name are basically the same.

also instead of hard coding "eosin.kvram"_n we should use eosin::kv_ram constant 
```suggestion
   template <typename T, eosio::name::raw SingletonName="anonymous"_n, eosio::name::raw DbName = eosio::kv_ram>
```

Reply:
You have to give the singleton a name.  Otherwise you can only have one singleton (one **total** not per-type) in your whole contract.

Reply:
You do give it a name if you want something other than the default... oh the name needs to include info from T.